File: vmtest.js
#!/usr/bin/env node

/**
 * VMTEST Environment Measurements - Node.js Version
 * Extracts timing, scheduling, cache, and memory measurements to detect virtualization
 * Based on research from Lin et al. (2021) and other VM detection papers
 */

const cluster = require('cluster');
const os = require('os');
const process = require('process');
const crypto = require('crypto');
const { performance } = require('perf_hooks');
const { Worker, isMainThread, parentPort, workerData } = require('worker_threads');

class VMTestMeasurements {
    constructor(iterations = 1000) {
        this.iterations = iterations;
        this.measurements = {};
        this.systemInfo = {};
    }

    // High-resolution timer in nanoseconds
    getTimeNs() {
        const hrTime = process.hrtime();
        return hrTime[0] * 1000000000 + hrTime[1];
    }

    // Statistical calculation methods
    static mean(values) {
        if (!values || values.length === 0) return 0;
        return values.reduce((sum, val) => sum + val, 0) / values.length;
    }

    static variance(values, meanVal = null) {
        if (!values || values.length < 2) return 0;
        if (meanVal === null) meanVal = VMTestMeasurements.mean(values);
        const sumSquaredDiffs = values.reduce((sum, val) => sum + Math.pow(val - meanVal, 2), 0);
        return sumSquaredDiffs / (values.length - 1);
    }

    static standardDeviation(values, meanVal = null) {
        return Math.sqrt(VMTestMeasurements.variance(values, meanVal));
    }

    static coefficientOfVariation(values) {
        const meanVal = VMTestMeasurements.mean(values);
        if (meanVal === 0) return 0;
        const stdDev = VMTestMeasurements.standardDeviation(values, meanVal);
        return stdDev / meanVal;
    }

    static skewness(values) {
        if (!values || values.length < 3) return 0;
        const meanVal = VMTestMeasurements.mean(values);
        const stdDev = VMTestMeasurements.standardDeviation(values, meanVal);
        if (stdDev <= 0) return 0;

        const sumCubedDeviations = values.reduce((sum, val) => {
            const deviation = (val - meanVal) / stdDev;
            return sum + Math.pow(deviation, 3);
        }, 0);

        let skew = sumCubedDeviations / values.length;

        // Apply bias correction for small samples
        if (values.length > 2) {
            skew = skew * Math.sqrt(values.length * (values.length - 1)) / (values.length - 2);
        }

        // Bound checking
        return Math.max(-100, Math.min(100, skew));
    }

    static kurtosis(values) {
        if (!values || values.length < 4) return 0;
        const meanVal = VMTestMeasurements.mean(values);
        const stdDev = VMTestMeasurements.standardDeviation(values, meanVal);
        if (stdDev <= 0) return 0;

        const sumQuartedDeviations = values.reduce((sum, val) => {
            const deviation = (val - meanVal) / stdDev;
            return sum + Math.pow(deviation, 4);
        }, 0);

        let kurt = (sumQuartedDeviations / values.length) - 3.0; // Excess kurtosis

        // Apply bias correction for small samples
        if (values.length > 3) {
            const n = values.length;
            kurt = ((n - 1) / ((n - 2) * (n - 3))) * ((n + 1) * kurt + 6);
        }

        // Bound checking
        return Math.max(-100, Math.min(100, kurt));
    }

    // CPU-bound computation for timing measurements
    cpuBoundTask() {
        const start = this.getTimeNs();
        
        // Computational work similar to C version
        let result = 0;
        for (let i = 0; i < 10000; i++) {
            result += Math.sqrt(i) * Math.sin(i) + Math.cos(i * 0.1);
        }
        
        const end = this.getTimeNs();
        return end - start;
    }

    // Memory-intensive task for cache testing
    cacheUnfriendlyAccess() {
        const start = this.getTimeNs();
        const size = 1024 * 1024; // 1MB
        const buffer = Buffer.alloc(size);
        
        // Random access pattern (cache-unfriendly)
        for (let i = 0; i < 1000; i++) {
            const offset = Math.floor(Math.random() * (size - 4));
            buffer.writeInt32BE(i, offset);
        }
        
        const end = this.getTimeNs();
        return end - start;
    }

    cacheFriendlyAccess() {
        const start = this.getTimeNs();
        const size = 1024 * 1024; // 1MB
        const buffer = Buffer.alloc(size);
        
        // Sequential access pattern (cache-friendly)
        for (let i = 0; i < size - 4; i += 4) {
            buffer.writeInt32BE(i, i);
        }
        
        const end = this.getTimeNs();
        return end - start;
    }

    // Memory address entropy calculation
    calculateMemoryEntropy() {
        const addresses = [];
        for (let i = 0; i < 1000; i++) {
            const buffer = Buffer.alloc(64);
            // Use buffer address as a proxy for memory address patterns
            addresses.push(buffer.byteOffset);
        }

        // Calculate Shannon entropy
        const addressCounts = {};
        addresses.forEach(addr => {
            const bucketAddr = Math.floor(addr / 4096); // 4KB buckets
            addressCounts[bucketAddr] = (addressCounts[bucketAddr] || 0) + 1;
        });

        let entropy = 0;
        const total = addresses.length;
        Object.values(addressCounts).forEach(count => {
            const probability = count / total;
            if (probability > 0) {
                entropy -= probability * Math.log2(probability);
            }
        });

        return entropy;
    }

    // Basic timing measurements
    async measureBasicTiming() {
        console.log("Measuring basic timing patterns...");
        const times = [];
        
        for (let i = 0; i < this.iterations; i++) {
            const execTime = this.cpuBoundTask();
            times.push(execTime);
        }

        const mean = VMTestMeasurements.mean(times);
        const variance = VMTestMeasurements.variance(times, mean);
        const cv = VMTestMeasurements.coefficientOfVariation(times);
        const skewness = VMTestMeasurements.skewness(times);
        const kurtosis = VMTestMeasurements.kurtosis(times);

        this.measurements.TIMING_BASIC_MEAN = mean;
        this.measurements.TIMING_BASIC_VARIANCE = variance;
        this.measurements.TIMING_BASIC_CV = cv;
        this.measurements.TIMING_BASIC_SKEWNESS = skewness;
        this.measurements.TIMING_BASIC_KURTOSIS = kurtosis;
    }

    // Consecutive timing measurements
    async measureConsecutiveTiming() {
        console.log("Measuring consecutive timing patterns...");
        const times = [];
        
        for (let i = 0; i < this.iterations; i++) {
            const start = this.getTimeNs();
            // Two consecutive operations
            this.cpuBoundTask();
            this.cpuBoundTask();
            const end = this.getTimeNs();
            times.push(end - start);
        }

        const mean = VMTestMeasurements.mean(times);
        const variance = VMTestMeasurements.variance(times, mean);
        const cv = VMTestMeasurements.coefficientOfVariation(times);
        const skewness = VMTestMeasurements.skewness(times);
        const kurtosis = VMTestMeasurements.kurtosis(times);

        this.measurements.TIMING_CONSECUTIVE_MEAN = mean;
        this.measurements.TIMING_CONSECUTIVE_VARIANCE = variance;
        this.measurements.TIMING_CONSECUTIVE_CV = cv;
        this.measurements.TIMING_CONSECUTIVE_SKEWNESS = skewness;
        this.measurements.TIMING_CONSECUTIVE_KURTOSIS = kurtosis;
    }

    // Thread scheduling measurements using Worker threads
    async measureThreadScheduling() {
        console.log("Measuring thread scheduling patterns...");
        
        return new Promise((resolve) => {
            const times = [];
            let completed = 0;
            const target = this.iterations/10;

            const workerCode = `
                const { parentPort, workerData } = require('worker_threads');
                const { performance } = require('perf_hooks');
                
                function cpuBoundTask() {
                    const start = performance.now();
                    let result = 0;
                    for (let i = 0; i < 10000; i++) {
                        result += Math.sqrt(i) * Math.sin(i) + Math.cos(i * 0.1);
                    }
                    const end = performance.now();
                    return (end - start) * 1000000; // Convert to nanoseconds
                }
                
                parentPort.postMessage(cpuBoundTask());
            `;

            for (let i = 0; i < target; i++) {
                const worker = new Worker(workerCode, { eval: true });
                
                worker.on('message', (time) => {
                    times.push(time);
                    completed++;
                    
                    if (completed === target) {
                        const mean = times.reduce((sum, val) => sum + val, 0) / times.length;
                        const variance = times.reduce((sum, val) => sum + Math.pow(val - mean, 2), 0) / (times.length - 1);
                        const stdDev = Math.sqrt(variance);
                        const cv = stdDev / mean;
                        
                        // Calculate skewness
                        const skewness = times.reduce((sum, val) => {
                            const deviation = (val - mean) / stdDev;
                            return sum + Math.pow(deviation, 3);
                        }, 0) / times.length;
                        
                        // Calculate kurtosis
                        const kurtosis = (times.reduce((sum, val) => {
                            const deviation = (val - mean) / stdDev;
                            return sum + Math.pow(deviation, 4);
                        }, 0) / times.length) - 3.0;

                        // Physical Machine Index (PMI) calculation
                        const pmi = (skewness * kurtosis) / (cv * 100)

                        this.measurements.SCHEDULING_THREAD_MEAN = mean;
                        this.measurements.SCHEDULING_THREAD_VARIANCE = variance;
                        this.measurements.SCHEDULING_THREAD_CV = cv;
                        this.measurements.SCHEDULING_THREAD_SKEWNESS = skewness;
                        this.measurements.SCHEDULING_THREAD_KURTOSIS = kurtosis;
                        this.measurements.PHYSICAL_MACHINE_INDEX = pmi;
                        
                        resolve();
                    }
                });
                
                worker.on('error', (err) => {
                    console.error('Worker error:', err);
                    completed++;
                    if (completed === target) resolve();
                });
            }
        });
    }

    // Multiprocessing measurements using child processes
    async measureMultiprocessing() {
        console.log("Measuring multiprocessing patterns...");
        const { spawn } = require('child_process');
        const times = [];

        return new Promise((resolve) => {
            let completed = 0;
            const target = Math.min(this.iterations, 100); // Limit for performance

            for (let i = 0; i < target; i++) {
                const start = this.getTimeNs();
                
                const child = spawn('node', ['-e', `
                    let result = 0;
                    for (let i = 0; i < 10000; i++) {
                        result += Math.sqrt(i) * Math.sin(i) + Math.cos(i * 0.1);
                    }
                    process.exit(0);
                `]);

                child.on('close', () => {
                    const end = this.getTimeNs();
                    times.push(end - start);
                    completed++;

                    if (completed === target) {
                        const mean = VMTestMeasurements.mean(times);
                        const variance = VMTestMeasurements.variance(times, mean);
                        const cv = VMTestMeasurements.coefficientOfVariation(times);
                        const skewness = VMTestMeasurements.skewness(times);
                        const kurtosis = VMTestMeasurements.kurtosis(times);

                        const pmi = (skewness * kurtosis) / (cv * 100);

                        this.measurements.SCHEDULING_MULTIPROC_MEAN = mean;
                        this.measurements.SCHEDULING_MULTIPROC_VARIANCE = variance;
                        this.measurements.SCHEDULING_MULTIPROC_CV = cv;
                        this.measurements.SCHEDULING_MULTIPROC_SKEWNESS = skewness;
                        this.measurements.SCHEDULING_MULTIPROC_KURTOSIS = kurtosis;
                        this.measurements.MULTIPROC_PHYSICAL_MACHINE_INDEX = pmi;

                        resolve();
                    }
                });
            }
        });
    }

    // Cache behavior measurements
    async measureCacheBehavior() {
        console.log("Measuring cache behavior patterns...");
        const unfriendlyTimes = [];
        const friendlyTimes = [];

        for (let i = 0; i < Math.min(this.iterations, 100); i++) {
            unfriendlyTimes.push(this.cacheUnfriendlyAccess());
            friendlyTimes.push(this.cacheFriendlyAccess());
        }

        const unfriendlyMean = VMTestMeasurements.mean(unfriendlyTimes);
        const friendlyMean = VMTestMeasurements.mean(friendlyTimes);

        this.measurements.CACHE_ACCESS_RATIO = unfriendlyMean / friendlyMean;
        this.measurements.CACHE_MISS_RATIO = (unfriendlyMean - friendlyMean) / friendlyMean;
    }

    // Memory entropy measurement
    async measureMemoryEntropy() {
        console.log("Measuring memory address entropy...");
        this.measurements.MEMORY_ADDRESS_ENTROPY = this.calculateMemoryEntropy();
    }

    // Gather system information
    gatherSystemInfo() {
        console.log("Gathering system information...");
        
        this.systemInfo = {
            platform: `${os.platform()} ${os.release()}`,
            hostname: os.hostname(),
            architecture: os.arch(),
            cpu_count: os.cpus().length,
            total_memory: os.totalmem(),
            free_memory: os.freemem(),
            node_version: process.version,
            cpu_model: os.cpus()[0]?.model || 'Unknown',
            load_average: os.loadavg(),
            uptime: os.uptime()
        };

        // Check for VM indicators
        const vmIndicators = {
            low_memory: this.systemInfo.total_memory < 2 * 1024 * 1024 * 1024, // < 2GB
            single_cpu: this.systemInfo.cpu_count === 1,
            vm_cpu_model: /virtual|qemu|kvm|xen|vmware/i.test(this.systemInfo.cpu_model)
        };

        this.systemInfo.vm_indicators = vmIndicators;
    }

    // Calculate overall VM likelihood
    calculateVMIndicators() {
        const indicators = {
            high_scheduling_variance: this.measurements.SCHEDULING_THREAD_CV > 0.15,
            low_pmi: this.measurements.PHYSICAL_MACHINE_INDEX < 1.0,
            high_timing_variance: this.measurements.TIMING_BASIC_CV > 0.1,
            abnormal_cache_ratio: this.measurements.CACHE_ACCESS_RATIO > 2.0,
            low_memory_entropy: this.measurements.MEMORY_ADDRESS_ENTROPY < 3.0
        };

        // Calculate likelihood score
        const weights = {
            high_scheduling_variance: 0.4,
            low_pmi: 0.3,
            high_timing_variance: 0.15,
            abnormal_cache_ratio: 0.1,
            low_memory_entropy: 0.05
        };

        let score = 0;
        Object.entries(indicators).forEach(([key, value]) => {
            if (value) score += weights[key] || 0;
        });

        indicators.vm_likelihood_score = score;
        indicators.likely_vm = score > 0.5;

        return indicators;
    }

    // Run all measurements
    async runAllMeasurements() {
        console.log("Starting VMtest measurements...");
        console.log(`Node.js version: ${process.version}`);
        console.log(`Platform: ${os.platform()} ${os.release()}`);
        console.log(`CPU count: ${os.cpus().length}`);
        console.log(`Memory: ${(os.totalmem() / 1024 / 1024 / 1024).toFixed(2)} GB`);
        console.log(`Iterations: ${this.iterations}`);
        console.log("");

        this.gatherSystemInfo();

        await this.measureBasicTiming();
        await this.measureConsecutiveTiming();
        await this.measureThreadScheduling();
        await this.measureMultiprocessing();
        await this.measureCacheBehavior();
        await this.measureMemoryEntropy();

        // Calculate overall coefficients of variation
        const timingCVs = [
            this.measurements.TIMING_BASIC_CV,
            this.measurements.TIMING_CONSECUTIVE_CV
        ];
        const schedulingCVs = [
            this.measurements.SCHEDULING_THREAD_CV,
            this.measurements.SCHEDULING_MULTIPROC_CV
        ];

        this.measurements.OVERALL_TIMING_CV = VMTestMeasurements.mean(timingCVs);
        this.measurements.OVERALL_SCHEDULING_CV = VMTestMeasurements.mean(schedulingCVs);

        const vmIndicators = this.calculateVMIndicators();

        return {
            system_info: this.systemInfo,
            measurements: this.measurements,
            vm_indicators: vmIndicators,
            timestamp: new Date().toISOString(),
            language: 'Node.js',
            version: '1.0.0'
        };
    }
}

// Main execution
async function main() {
    const iterations = process.argv[2] ? parseInt(process.argv[2]) : 1000;
    
    if (isNaN(iterations) || iterations < 1) {
        console.error("Usage: node vmtest.js [iterations]");
        console.error("Example: node vmtest.js 1000");
        process.exit(1);
    }

    const vmtest = new VMTestMeasurements(iterations);
    
    try {
        const results = await vmtest.runAllMeasurements();
        
        console.log("\n" + "=".repeat(50));
        console.log("VMTEST RESULTS");
        console.log("=".repeat(50));
        
        // Print formatted results
        console.log(JSON.stringify(results, null, 2));
        
        console.log("\n" + "=".repeat(50));
        console.log("VM DETECTION SUMMARY");
        console.log("=".repeat(50));
        
        const { vm_indicators } = results;
        console.log(`VM Likelihood Score: ${(vm_indicators.vm_likelihood_score * 100).toFixed(1)}%`);
        console.log(`Likely Virtual Machine: ${vm_indicators.likely_vm ? 'YES' : 'NO'}`);
        
        if (vm_indicators.likely_vm) {
            console.log("\nVM Indicators Detected:");
            Object.entries(vm_indicators).forEach(([key, value]) => {
                if (value === true && key !== 'likely_vm') {
                    console.log(`  - ${key.replace(/_/g, ' ')}`);
                }
            });
        }
        
    } catch (error) {
        console.error("Error running VMtest:", error);
        process.exit(1);
    }
}

// Only run main if this file is executed directly
if (require.main === module) {
    main().catch(console.error);
}

module.exports = VMTestMeasurements;

File: vmtest.rb
#!/usr/bin/env ruby
class Array
  def sum(initial = 0)
    inject(initial, :+)
  end unless method_defined?(:sum)
end
# VMtest - System Measurements Tool (Pure Measurements Only)
# Extracts timing, scheduling, cache, and memory measurements
# Based on research from Lin et al. (2021) and other academic sources
# 
# This version collects only raw measurements without VM detection logic

require 'json'
require 'thread'
require 'etc'
require 'rbconfig'

class VMTest
  THREAD_COUNT = 4
  CACHE_SIZE = 1024 * 1024  # 1MB

  def initialize(iterations = 1000)
    @iterations = iterations
    @measurements = {}
    @system_info = {}
  end

  # Get high-resolution time in nanoseconds
  def get_time_ns
    Process.clock_gettime(Process::CLOCK_MONOTONIC, :nanosecond)
  end

  # Statistical calculation methods
  def self.mean(values)
    return 0.0 if values.nil? || values.empty?
    values.sum.to_f / values.length
  end

  def self.variance(values)
    return 0.0 if values.nil? || values.length <= 1
    
    mean_val = mean(values)
    sum_squared_diff = values.map { |x| (x - mean_val) ** 2 }.sum
    
    sum_squared_diff / (values.length - 1)
  end

  def self.standard_deviation(values)
    Math.sqrt(variance(values))
  end

  def self.coefficient_of_variation(values)
    return 0.0 if values.nil? || values.empty?
    
    mean_val = mean(values)
    return 0.0 if mean_val == 0
    
    std_dev = standard_deviation(values)
    std_dev / mean_val
  end

  def self.skewness(values)
    return 0.0 if values.nil? || values.length < 3
    
    n = values.length
    mean_val = mean(values)
    variance_val = variance(values)
    
    return 0.0 if variance_val <= 0
    
    std_dev = Math.sqrt(variance_val)
    
    # Calculate third moment about the mean
    m3 = values.map { |x| ((x - mean_val) / std_dev) ** 3 }.sum / n
    
    # Apply bias correction for sample skewness
    if n > 2
      adjustment = Math.sqrt(n * (n - 1)) / (n - 2)
      m3 * adjustment
    else
      m3
    end
  end

  def self.kurtosis(values)
    return 0.0 if values.nil? || values.length < 4
    
    n = values.length
    mean_val = mean(values)
    variance_val = variance(values)
    
    return 0.0 if variance_val <= 0
    
    std_dev = Math.sqrt(variance_val)
    
    # Calculate fourth moment about the mean
    m4 = values.map { |x| ((x - mean_val) / std_dev) ** 4 }.sum / n
    
    # Excess kurtosis (biased)
    kurt_biased = m4 - 3.0
    
    # Apply bias correction for sample kurtosis
    if n > 3
      ((n - 1) * ((n + 1) * kurt_biased + 6)) / ((n - 2) * (n - 3))
    else
      kurt_biased
    end
  end

  # Calculate raw Physical Machine Index (no logarithm)
  def calculate_raw_pmi(kurtosis, skewness, variance)
    return 0.0 if variance <= 0
    (kurtosis * skewness) / variance
  end

  # Calculate Shannon entropy
  def calculate_entropy(values)
    return 0.0 if values.nil? || values.empty?
    
    min_val = values.min
    max_val = values.max
    
    return 0.0 if min_val == max_val
    
    # Create histogram with 20 bins
    bins = 20
    bin_width = (max_val - min_val).to_f / bins
    histogram = Array.new(bins, 0)
    
    # Fill histogram
    values.each do |val|
      bin_idx = ((val - min_val) / bin_width).to_i
      bin_idx = bins - 1 if bin_idx >= bins
      histogram[bin_idx] += 1
    end
    
    # Calculate entropy
    entropy = 0.0
    total = values.length
    histogram.each do |count|
      if count > 0
        probability = count.to_f / total
        entropy -= probability * Math.log2(probability)
      end
    end
    
    entropy
  end

  # Gather system information
  def gather_system_info
    puts "Gathering system information..."
    
    @system_info = {
      platform: "#{RbConfig::CONFIG['host_os']} #{`uname -r`.chomp}",
      hostname: `hostname`.chomp,
      architecture: RbConfig::CONFIG['host_cpu'],
      cpu_count: Etc.nprocessors,
      ruby_version: RUBY_VERSION,
      ruby_platform: RUBY_PLATFORM,
      ruby_engine: RUBY_ENGINE,
      timestamp: Time.now.to_f
    }
    
    # Try to get memory info
    begin
      if File.exist?('/proc/meminfo')
        File.readlines('/proc/meminfo').each do |line|
          if line.start_with?('MemTotal:')
            # Convert KB to bytes
            @system_info[:total_memory] = line.split[1].to_i * 1024
            break
          end
        end
      end
    rescue
      # Ignore errors
    end
    
    # Try to get CPU frequency
    begin
      if File.exist?('/proc/cpuinfo')
        File.readlines('/proc/cpuinfo').each do |line|
          if line.start_with?('cpu MHz')
            @system_info[:cpu_freq] = line.split(':')[1].strip.to_f
            break
          end
        end
      end
    rescue
      # Ignore errors
    end
  end

  # CPU-bound computation for timing measurements
  def cpu_bound_task
    start_time = get_time_ns
    
    # Computational work similar to other implementations
    result = 0
    10000.times do |i|
      result += i * i
    end
    
    end_time = get_time_ns
    end_time - start_time
  end

  # Basic timing measurements
  def measure_basic_timing
    puts "Measuring basic timing patterns..."
    
    times = []
    
    @iterations.times do
      times << cpu_bound_task
    end
    
    @measurements[:TIMING_BASIC_MEAN] = self.class.mean(times)
    @measurements[:TIMING_BASIC_VARIANCE] = self.class.variance(times)
    @measurements[:TIMING_BASIC_CV] = self.class.coefficient_of_variation(times)
    @measurements[:TIMING_BASIC_SKEWNESS] = self.class.skewness(times)
    @measurements[:TIMING_BASIC_KURTOSIS] = self.class.kurtosis(times)
  end

  # Consecutive timing measurements
  def measure_consecutive_timing
    puts "Measuring consecutive timing patterns..."
    
    times = []
    
    @iterations.times do
      start_time = get_time_ns
      
      # Two consecutive operations
      result1 = 0
      10000.times { |i| result1 += i * i }
      result2 = 0
      10000.times { |i| result2 += i * i }
      
      end_time = get_time_ns
      times << (end_time - start_time)
    end
    
    @measurements[:TIMING_CONSECUTIVE_MEAN] = self.class.mean(times)
    @measurements[:TIMING_CONSECUTIVE_VARIANCE] = self.class.variance(times)
    @measurements[:TIMING_CONSECUTIVE_CV] = self.class.coefficient_of_variation(times)
    @measurements[:TIMING_CONSECUTIVE_SKEWNESS] = self.class.skewness(times)
    @measurements[:TIMING_CONSECUTIVE_KURTOSIS] = self.class.kurtosis(times)
  end

  # Thread scheduling measurements
  def measure_thread_scheduling
    puts "Measuring thread scheduling patterns..."
    
    times = []
    completed = 0
    mutex = Mutex.new
    target = [@iterations, 500].min  # Limit for performance
    
    # Use a queue to manage work
    work_queue = Queue.new
    target.times { work_queue << true }

    # Worker thread function
    worker = proc do
      while work_queue.length > 0
        begin
          work_queue.pop(true)  # non-blocking pop
          exec_time = cpu_bound_task
          
          mutex.synchronize do
            times << exec_time
            completed += 1
          end
        rescue ThreadError
          # Queue is empty
          break
        end
      end
    end

    start_time = get_time_ns

    # Create and start worker threads
    threads = []
    THREAD_COUNT.times do
      threads << Thread.new(&worker)
    end

    # Wait for all threads to complete
    threads.each(&:join)

    end_time = get_time_ns

    # Calculate statistics using thread execution times
    thread_times = [end_time - start_time]  # Overall thread execution time
    target.times do |i|
      thread_times << times[i] if times[i]
    end

    @measurements[:SCHEDULING_THREAD_MEAN] = self.class.mean(thread_times)
    @measurements[:SCHEDULING_THREAD_VARIANCE] = self.class.variance(thread_times)
    @measurements[:SCHEDULING_THREAD_CV] = self.class.coefficient_of_variation(thread_times)
    @measurements[:SCHEDULING_THREAD_SKEWNESS] = self.class.skewness(thread_times)
    @measurements[:SCHEDULING_THREAD_KURTOSIS] = self.class.kurtosis(thread_times)

    # CORRECTED PMI calculation using raw formula (no logarithm)
    @measurements[:PHYSICAL_MACHINE_INDEX] = calculate_raw_pmi(
      @measurements[:SCHEDULING_THREAD_KURTOSIS],
      @measurements[:SCHEDULING_THREAD_SKEWNESS],
      @measurements[:SCHEDULING_THREAD_VARIANCE]
    )
  end

  # Multiprocessing measurements using fork
  def measure_multiprocessing
    puts "Measuring multiprocessing patterns..."
    
    times = []
    target = [@iterations, 100].min  # Limit for performance

    target.times do
      start_time = get_time_ns
      
      # Create multiple child processes
      pids = []
      4.times do
        pid = fork do
          # CPU-bound work in child process
          result = 0
          10000.times do |i|
            result += i * i
          end
          exit(0)
        end
        pids << pid if pid
      end
      
      # Wait for all children
      pids.each { |pid| Process.wait(pid) }
      
      end_time = get_time_ns
      times << (end_time - start_time)
    end

    @measurements[:SCHEDULING_MULTIPROC_MEAN] = self.class.mean(times)
    @measurements[:SCHEDULING_MULTIPROC_VARIANCE] = self.class.variance(times)
    @measurements[:SCHEDULING_MULTIPROC_CV] = self.class.coefficient_of_variation(times)
    @measurements[:SCHEDULING_MULTIPROC_SKEWNESS] = self.class.skewness(times)
    @measurements[:SCHEDULING_MULTIPROC_KURTOSIS] = self.class.kurtosis(times)

    # Calculate raw PMI for multiprocessing (no logarithm)
    @measurements[:MULTIPROC_PHYSICAL_MACHINE_INDEX] = calculate_raw_pmi(
      @measurements[:SCHEDULING_MULTIPROC_KURTOSIS],
      @measurements[:SCHEDULING_MULTIPROC_SKEWNESS],
      @measurements[:SCHEDULING_MULTIPROC_VARIANCE]
    )
  end

  # Cache behavior measurements
  def measure_cache_behavior
    puts "Measuring cache behavior patterns..."
    
    # Create large data array
    data = Array.new(CACHE_SIZE) { rand }
    
    # Cache-friendly access (sequential)
    cache_friendly_times = []
    [@iterations, 100].min.times do
      start_time = get_time_ns
      
      sum = 0
      (0...CACHE_SIZE).step(1000) do |i|
        sum += data[i]
      end
      
      end_time = get_time_ns
      cache_friendly_times << (end_time - start_time)
    end
    
    # Cache-unfriendly access (random)
    indices = (0...CACHE_SIZE).to_a.shuffle
    cache_unfriendly_times = []
    
    [@iterations, 100].min.times do
      start_time = get_time_ns
      
      sum = 0
      (0...CACHE_SIZE).step(1000) do |i|
        sum += data[indices[i]]
      end
      
      end_time = get_time_ns
      cache_unfriendly_times << (end_time - start_time)
    end
    
    # Calculate cache metrics
    cache_friendly_mean = self.class.mean(cache_friendly_times)
    cache_unfriendly_mean = self.class.mean(cache_unfriendly_times)
    
    if cache_friendly_mean > 0
      @measurements[:CACHE_ACCESS_RATIO] = cache_unfriendly_mean / cache_friendly_mean
      @measurements[:CACHE_MISS_RATIO] = (cache_unfriendly_mean - cache_friendly_mean) / cache_friendly_mean
    else
      @measurements[:CACHE_ACCESS_RATIO] = 1.0
      @measurements[:CACHE_MISS_RATIO] = 0.0
    end
  end

  # Memory address entropy calculation
  def measure_memory_entropy
    puts "Measuring memory entropy patterns..."
    
    begin
      # Method 1: Use allocation timing patterns as entropy proxy
      timings = []
      objects = []
      
      1000.times do |i|
        start_time = get_time_ns
        
        # Allocate memory and use object_id as address proxy
        size = 64 + (i * 16)
        buffer = " " * size
        objects << buffer
        
        # Touch the memory
        buffer[0] = (i % 256).chr
        
        end_time = get_time_ns
        timings << (end_time - start_time)
      end
      
      # Calculate entropy of allocation timings
      if timings.length > 0
        entropy = calculate_entropy(timings)
        @measurements[:MEMORY_ADDRESS_ENTROPY] = entropy
      else
        @measurements[:MEMORY_ADDRESS_ENTROPY] = 0.0
      end
      
      # If entropy is very low, try alternative method
      if @measurements[:MEMORY_ADDRESS_ENTROPY] < 0.1
        # Method 2: Use object_id patterns
        object_ids = []
        100.times do
          buffer = " " * 64
          object_ids << buffer.object_id
        end
        
        # Calculate entropy using object_id differences
        if object_ids.length > 1
          diffs = []
          (object_ids.length - 1).times do |i|
            diffs << (object_ids[i + 1] - object_ids[i]).abs
          end
          
          if diffs.length > 0
            entropy = calculate_entropy(diffs)
            @measurements[:MEMORY_ADDRESS_ENTROPY] = [entropy, 0.1].max
          end
        end
      end
      
      # Final fallback: system-based estimation
      if @measurements[:MEMORY_ADDRESS_ENTROPY] < 0.1
        cpu_count = Etc.nprocessors
        # Estimate entropy based on system characteristics
        entropy = Math.log2(cpu_count) + 2.0  # Base entropy estimate
        @measurements[:MEMORY_ADDRESS_ENTROPY] = [[entropy, 1.0].max, 4.0].min
      end
      
    rescue => e
      puts "Error in memory entropy measurement: #{e}"
      @measurements[:MEMORY_ADDRESS_ENTROPY] = 1.0  # Default reasonable value
    end
  end

  # Calculate composite measurements
  def calculate_composite_measurements
    puts "Calculating composite measurements..."
    
    # Overall timing CV
    timing_cvs = []
    timing_cvs << @measurements[:TIMING_BASIC_CV] if @measurements[:TIMING_BASIC_CV] > 0
    timing_cvs << @measurements[:TIMING_CONSECUTIVE_CV] if @measurements[:TIMING_CONSECUTIVE_CV] > 0
    
    @measurements[:OVERALL_TIMING_CV] = timing_cvs.empty? ? 0.0 : self.class.mean(timing_cvs)
    
    # Overall scheduling CV
    scheduling_cvs = []
    scheduling_cvs << @measurements[:SCHEDULING_THREAD_CV] if @measurements[:SCHEDULING_THREAD_CV] > 0
    scheduling_cvs << @measurements[:SCHEDULING_MULTIPROC_CV] if @measurements[:SCHEDULING_MULTIPROC_CV] > 0
    
    @measurements[:OVERALL_SCHEDULING_CV] = scheduling_cvs.empty? ? 0.0 : self.class.mean(scheduling_cvs)
  end

  # Run all measurements
  def run_all_measurements
    puts "Starting system measurements..."
    puts "Ruby version: #{RUBY_VERSION}"
    puts "Platform: #{RbConfig::CONFIG['host_os']}"
    puts "CPU count: #{Etc.nprocessors}"
    
    begin
      if @system_info[:total_memory]
        puts "Memory: #{(@system_info[:total_memory] / 1024.0 / 1024.0 / 1024.0).round(2)} GB"
      end
    rescue
      # Memory info not available
    end
    
    puts "Iterations: #{@iterations}"
    puts ""

    gather_system_info
    measure_basic_timing
    measure_consecutive_timing
    measure_thread_scheduling
    
    # Only measure multiprocessing on Unix-like systems
    if RbConfig::CONFIG['host_os'] !~ /mswin|mingw|cygwin/i
      measure_multiprocessing
    else
      puts "Skipping multiprocessing measurements on Windows"
      @measurements[:SCHEDULING_MULTIPROC_MEAN] = 0.0
      @measurements[:SCHEDULING_MULTIPROC_VARIANCE] = 0.0
      @measurements[:SCHEDULING_MULTIPROC_CV] = 0.0
      @measurements[:SCHEDULING_MULTIPROC_SKEWNESS] = 0.0
      @measurements[:SCHEDULING_MULTIPROC_KURTOSIS] = 0.0
      @measurements[:MULTIPROC_PHYSICAL_MACHINE_INDEX] = 0.0
    end
    
    measure_cache_behavior
    measure_memory_entropy
    calculate_composite_measurements

    puts "\nMeasurements complete!"
  end

  # Get results
  def get_results
    {
      system_info: @system_info,
      measurements: @measurements,
      timestamp: Time.now.strftime('%Y-%m-%dT%H:%M:%S%z'),
      language: 'ruby',
      version: '1.0.0'
    }
  end

  # Print results in JSON format
  def print_results_json
    results = get_results
    puts JSON.pretty_generate(results)
  end

  # Save results to file
  def save_results_json(filename = nil)
    filename ||= "measurements_#{Time.now.strftime('%Y%m%d_%H%M%S')}.json"
    
    File.write(filename, JSON.pretty_generate(get_results))
    puts "Measurements saved to: #{filename}"
    filename
  end
end

# Main function
def main
  # Parse command line arguments
  iterations = 1000
  if ARGV.length > 0
    begin
      arg = ARGV[0].to_i
      iterations = arg if arg > 0
    rescue
      puts "Invalid iterations argument, using default 1000"
    end
  end

  # Create and run VMTest
  vmtest = VMTest.new(iterations)
  vmtest.run_all_measurements
  
  # Output results
  puts "\nResults:"
  puts "=" * 50
  vmtest.print_results_json
end

# Run main function if this file is executed directly
if __FILE__ == $0
  main
end
